R/fileupload.R

```diff
@@ -20,6 +20,104 @@
+# https://msdn.microsoft.com/en-us/library/aa365247.aspx
+# https://blogs.msdn.microsoft.com/oldnewthing/20031022-00/?p=42073
+illegalWindowsNames <-
+  c("AUX",
+    paste0("COM", 1:9),
+    "CON",
+    paste0("LPT", 1:9),
+    "NUL",
+    "PRN")
+
+# @details Removes strings in file names that could cause problems on Windows.
+# @param str A character vector to remove illegal strings from.
+# @param illegal The vector of words, by themselves or with an extension, to
+#   remove from \code{str}
+# @return \code{str} with any illegal words removed.
+removeIllegalWindowsFilenames <- function(str, illegal) {
+  ret <- str
+  ret[ret %in% illegal] <- ""
+  illegalNames <- tools::file_path_sans_ext(ret) %in% illegal
+  ret[illegalNames] <- paste0(".", tools::file_ext(ret[illegalNames]))
+  ret
+}
+
+# @details Helper function for abbreviating dots and removing unwanted
+#   characters and strings from a filename.
+# @param str A character vector which is presumably the name of a file.
+# @param whitelist The regular expression fragment indicating the characters to
+#   keep.
+# @param windows A logical indicating whether or not to strip names that are
+#   illegal on Windows.
+# @return str with dots abbreviated, non-whitelisted characters removed, and
+#   optionally illegal Windows names removed.
+sanitize <- function(str, whitelist = "a-zA-Z0-9\\.", windows = FALSE) {
+  sanitized <- str %>%
+    gsub(sprintf("[^%s]", whitelist), "", .) %>%
+    gsub("\\.+", "\\.", .)
+
+  if(windows)
+    removeIllegalWindowsFilenames(sanitized, illegalWindowsNames)
+  else
+    sanitized
+}
+
+# @details \code{sanitizeFileName} is an effort to safely retain something close
+#   to an uploaded file's original name in order to support libraries like
+#   readxl that are sensitive to the names of files. A good overview of the
+#   risks associated with uploaded files is
+#   \url{https://www.owasp.org/index.php/Unrestricted_File_Upload}
+# @param name A character vector, the name of a file to sanitize.
+# @param default Character vector to return if \code{name} cannot be sanitized.
+#   If a file's extension can be preserved, and the total length of
+#   \code{default} and the extension is not more than \code{maxSize}, then
+#   \code{default} is combined with the extension and returned.
+# @param maxSize The maximum allowable size of the name returned by this
+#   function. (Default is 255 characters.)
+# @return The sanitized \code{name}, or the \code{default} combined with the
+#   original extension, if possible. Otherwise, \code{default}.
+sanitizeFileName <- function(name, default, maxSize = 255) {
+  if (missing(default))
+    stop("default is a required argument")
+
+  if (any(nchar(default) > maxSize))
+    stop("default can't be longer than maxSize")
+
+  if (length(name) != length(default))
+    stop("name and default must be the same length")
+
+  sanitizedNames <- sanitize(name, windows = (.Platform$OS.type == "windows"))
```

Comment:
It turns out we already have a function `isWindows()` in Shiny: https://github.com/rstudio/shiny/blob/852c000/R/utils.R#L310